swarm/storage/mru/handler.go

```diff
@@ -370,20 +369,19 @@ func (h *Handler) Load(rootAddr storage.Address) (*resource, error) {
 		return nil, NewError(ErrNotFound, err.Error())
 	}
 
-	// \TODO this is not enough to make sure the data isn't bogus. A normal content addressed chunk could still satisfy these criteria
-	if len(chunk.SData) <= metadataChunkOffsetSize {
-		return nil, NewError(ErrNothingToReturn, fmt.Sprintf("Invalid chunk length %d, should be minimum %d", len(chunk.SData), metadataChunkOffsetSize+1))
-	}
-
 	// create the index entry
 	rsrc := &resource{}
-	rsrc.unmarshalBinary(chunk.SData)
+
+	if err := rsrc.resourceMetadata.binaryGet(chunk.SData); err != nil { // Will fail if this is not really a metadata chunk
```

Comment:
yes... We can rename it, but let's discuss serialization in the last PR. My reasoning is that normally a `Marshal` / `Unmarshal` method takes a byte array or return a byte array. `Marshal` methods return a byte array that then you have to copy to include it as part of the parent structure you are serializing. This is inefficient.
The idea of `binaryGet` and `binaryPut` is that they operate on bounded subslices, meaning that they must receive a slice of the exact size for serialization/deserialization otherwise they will return an error. This helps with unit testing.

`binaryGet` == `Unmarshal`
`binaryPut` != `Marshal` . You must provide a slice of exact size in which the serialized content will be put. Similar to e.g.,  `binary.LittleEndian.PutUint64(b,i)` needs an 8 byte slice in `b`.

I don't care about the exact naming of this, we can change it very quickly. I just would like to make sure I get my point across.

Reply:
I like this idea. But I can't remember seeing this approach used in the go-ethereum codebase. Maybe we should discuss this at a roundtable.

Reply:
Always open for innovation ;-)

Reply:
Yeah. I mean, it's kind of a C-bastard style way, this, and in a way quite un-go. That is to say; golang seems to have defined Marshal'ing as the way to do these operations. Part of me doesn't want to confuse by breaking convention. The other part of me is otoh quite adverse to such conformity-pressure.

Reply:
I would say C-badass style way ;-)

Anyway, you guys decide. If I change it, it'll be `Marshal()` and then `copy()` and that sucks to me, but if it is the code style, o well.

But there are go methods that put data on an existing slice, like the `binary` package, so I would argue that.

And I like this concept, because the caller must produce a slice that is *exactly* the required length, or else we won't accept it. This is very helpful to catch bugs early.

But let's `defer` on serialization until we look at it holistically in the last step.

Reply:
Badass, for sure. Make no mistake, it's close to my heart, too. My hyphen was misplaced.